daemon/access.go

```diff
@@ -0,0 +1,163 @@
+	var flags polkit.CheckFlags
+	allowHeader := r.Header.Get(client.AllowInteractionHeader)
+	if allowHeader != "" {
+		if allow, err := strconv.ParseBool(allowHeader); err != nil {
+			logger.Noticef("error parsing %s header: %s", client.AllowInteractionHeader, err)
+		} else if allow {
+			flags |= polkit.CheckAllowInteraction
+		}
+	}
+	// Pass both pid and uid from the peer ucred to avoid pid race
+	switch authorized, err := polkitCheckAuthorization(ucred.pid, ucred.uid, action, nil, flags); err {
+	case nil:
+		if authorized {
+			// polkit says user is authorised
+			return accessOK
+		}
+	case polkit.ErrDismissed:
+		return accessCancelled
+	default:
+		logger.Noticef("polkit error: %s", err)
+	}
+	return accessUnauthorized
+}
+
+// accessChecker checks whether a particular request is allowed.
+//
+// An access checker will either allow a request, deny it, or return
+// accessUnknown, which indicates the decision should be delegated to
+// the next access checker.
+type accessChecker interface {
+	checkAccess(r *http.Request, ucred *ucrednet, user *auth.UserState) accessResult
+}
+
+// requireSnapdSocket ensures the request was received via snapd.socket.
+func requireSnapdSocket(ucred *ucrednet) accessResult {
+	if ucred == nil {
+		return accessForbidden
+	}
+
+	if ucred.socket != dirs.SnapdSocket {
+		return accessForbidden
+	}
+
+	return accessOK
+}
+
+// openAccess allows requests without authentication, provided they
+// have peer credentials and were not received on snapd-snap.socket
+type openAccess struct{}
+
+func (ac openAccess) checkAccess(r *http.Request, ucred *ucrednet, user *auth.UserState) accessResult {
+	return requireSnapdSocket(ucred)
+}
+
+// authenticatedAccess allows requests from authenticated users,
+// provided they were not received on snapd-snap.socket
+//
+// A user is considered authenticated if they provide a macaroon, are
+// the root user according to peer credentials, or granted access by
+// Polkit.
+type authenticatedAccess struct {
+	Polkit string
+}
+
+func (ac authenticatedAccess) checkAccess(r *http.Request, ucred *ucrednet, user *auth.UserState) accessResult {
+	if result := requireSnapdSocket(ucred); result != accessOK {
+		return result
+	}
+
+	if user != nil {
+		return accessOK
+	}
+
+	if ucred.uid == 0 {
+		return accessOK
+	}
+
+	// We check polkit last because it may result in the user
+	// being prompted for authorisation.  This should be avoided
+	// if access is otherwise granted.
+	if ac.Polkit != "" {
+		return checkPolkitAction(r, ucred, ac.Polkit)
+	}
+
+	return accessUnauthorized
+}
+
+// rootAccess allows requests from the root uid, provided they
+// were not received on snapd-snap.socket
+type rootAccess struct{}
+
+func (ac rootAccess) checkAccess(r *http.Request, ucred *ucrednet, user *auth.UserState) accessResult {
+	if result := requireSnapdSocket(ucred); result != accessOK {
+		return result
+	}
+
+	if ucred.uid == 0 {
+		return accessOK
+	}
+	return accessForbidden
```

Comment:
this is an interesting case because while a token doesn't help switching to sudo would, basically it's uncelar that distinguishing only between Forbidden vs Authorize is enough to produce a good error above

Reply:
The HTTP error codes are a bit misnamed.  The "401 Unauthorized" error code is described as meaning authentication is required:

https://tools.ietf.org/html/rfc2616#section-10.4.2

More specifically, the status code should be accompanied by a `WWW-Authenticate` header explaining how the client can construct an appropriate `Authorization` header to authenticate a future request.

In contrast, "403 Forbidden" means that the server is refusing to carry out the request and that an `Authorization` header won't help.  In any other system, this would be called "Unauthorized".

For endpoints that are relying on peer credentials, we've authenticated the client and decided that there is nothing that particular client could do to have the request accepted.  I agree that we still want to include a hint in 403 responses that a request by root will be handled.